### PR TITLE
support for Webpush with VAPID

### PR DESCRIPTION
**Pull request by jkraemer**
This PR implements Webpush support. Webpush is supported by all major desktop browsers except Safari, and has by now replaced the vendor specific Mozilla / Google Chrome desktop push solutions. All the heavy lifting is done by the [webpush gem](https://github.com/zaru/webpush), so the implementation is pretty straightforward.

### CODE REVIEW

**Comment by benlangfeld** [comment]
Thanks for the contribution. I apologise for the delay, but I'll get you a review on this early next week.

File: README.md [hunk 1]
```diff
@@ -297,6 +298,49 @@ n.save!
...
+app = Rpush::Webpush::App.new
+app.name = 'webpush'
+app.certificate = vapid_keypair.merge(subject: 'user@example.org').to_json
+app.connections = 1
+app.save!
+```
+
+The `subscription` object you obtain from a subscribed browser holds an
+endpoint URL and cryptographic keys. When sending a notification, simply pass
+the whole subscription as sole member of the `endpoint_ids` collection:
```
Comment (benlangfeld) [inline README.md:329]: Did you mean this?

```suggestion
the whole subscription as sole member of the `registration_ids` collection:
```

File: lib/rpush/client/active_record/webpush/app.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+      module Webpush
+        class App < Rpush::Client::ActiveRecord::App
+          include Rpush::Client::ActiveModel::Webpush::App
+        end
+      end
+    end
+  end
+end
+
+
```
Comment (benlangfeld) [inline lib/rpush/client/active_record/webpush/app.rb:13]:
```suggestion
```

File: lib/rpush/client/redis/webpush/app.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+module Rpush
+  module Client
+    module Redis
+      module Webpush
+        class App < Rpush::Client::Redis::App
+          include Rpush::Client::ActiveModel::Webpush::App
+
+          def vapid_keypair=(value)
```
Comment (benlangfeld) [inline lib/rpush/client/redis/webpush/app.rb:8]: How come this isn't required for active_record?

File: lib/rpush/client/redis/webpush/notification.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+module Rpush
+  module Client
+    module Redis
+      module Webpush
+        class Notification < Rpush::Client::Redis::Notification
+          include Rpush::Client::ActiveModel::Webpush::Notification
+
+          def time_to_live=(value)
```
Comment (benlangfeld) [inline lib/rpush/client/redis/webpush/notification.rb:8]: How come this isn't required for active_record?

File: lib/rpush/client/redis/webpush/app.rb (hunk 3 above)
Comment (jkraemer) [reply lib/rpush/client/redis/webpush/app.rb:8]: I don't know, but without this, the Redis tests fail. The Pushy code has the same workaround. It looks like for AR, the `alias_attribute` call from the common module creates both reader and writer, while with Redis, it fails to create the writer method.

File: lib/rpush/client/redis/webpush/notification.rb (hunk 4 above)
Comment (jkraemer) [reply lib/rpush/client/redis/webpush/notification.rb:8]: same reason as with app.rb

File: README.md (hunk 1 above)
Comment (jkraemer) [reply README.md]: exactly :)

File: lib/rpush/client/active_model/webpush/app.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+          def service_name
+            'webpush'
+          end
+
+          def vapid
+            unless defined?(@vapid)
+              @vapid = JSON.parse(vapid_keypair).symbolize_keys
+            end
+            @vapid
```
Comment (benlangfeld) [inline lib/rpush/client/active_model/webpush/app.rb:37]: Any reason not to simply do this?

```suggestion
            @vapid ||= JSON.parse(vapid_keypair).symbolize_keys
```

File: lib/rpush/client/active_model/webpush/notification.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+          class RegistrationValidator < ::ActiveModel::Validator
+            KEYS = %i[ endpoint keys ].freeze
+            def validate(record)
+              return if record.registration_ids.blank?
+              return if record.registration_ids.size > 1
+              reg = record.registration_ids.first
+              unless reg.is_a?(Hash) and
+                  reg.keys.sort == KEYS and
+                  reg[:endpoint].is_a?(String) and
+                  reg[:keys].is_a?(Hash)
```
Comment (benlangfeld) [inline lib/rpush/client/active_model/webpush/notification.rb:16]:
```suggestion
              unless reg.is_a?(Hash) &&
                  reg.keys.sort == KEYS &&
                  reg[:endpoint].is_a?(String) &&
                  reg[:keys].is_a?(Hash)
```

**Review by benlangfeld — COMMENTED** [review]
Just a couple of minor comments but looks good otherwise.

**Review by benlangfeld — APPROVED** [review]

**Comment by jkraemer** [comment]
Thanks!

**Pull request merged**